contracts/contracts/keeper/Keeper.sol

```diff
@@ -0,0 +1,194 @@
+        string param;
+    }
+
+    struct RebaseInfo {
+        bool run;
+        string param;
+    }
+
+    struct DummyInfo {
+        bool rebase;
+        bool allocate;
+        uint256 time;
+    }
+
+    event KeeperEvent(string action, bool success, bool rebase, bool allocate);
+
+    // /*
+    //  * @notice modifier that allows it to be simulated via eth_call by checking
+    //  * that the sender is the zero address.
+    //  */
+    // modifier cannotExecute()
+    // {
+    //   preventExecution();
+    //   _;
+    // }
+
+    //  /*
+    //  * @notice method that allows it to be simulated via eth_call by checking that
+    //  * the sender is the zero address.
+    //  */
+    // function preventExecution() internal view {
+    //   console.log('In preventExecution');
+    //   console.log(tx.origin);
+    //   require(tx.origin == address(0), "only for simulated backend");
+    // }
+
+    constructor(address payable _vaultCoreAddr, address payable _ousdAddr)
+        public
+    {
+        vaultCore = VaultCore(_vaultCoreAddr);
+        oUSD = OUSD(_ousdAddr);
+        nextAllocate = now + 1 days;
+    }
+
+    /*
+     * @notice method that is simulated by the keepers to see if any work actually
+     * needs to be performed. This method does does not actually need to be
+     * executable, and since it is only ever simulated it can consume lots of gas.
+     * @dev To ensure that it is never called, you may want to add the
+     * cannotExecute modifier from KeeperBase to your implementation of this
+     * method.
+     * @return success boolean to indicate whether the keeper should call
+     * performUpkeep or not.
+     * @return success bytes that the keeper should call performUpkeep with, if
+     * upkeep is needed.
+     */
+
+    function checkUpkeep(bytes calldata _data)
+        external
+        returns (
+            // view
+            //cannotExecute
+            bool success,
+            bytes memory dynamicData
+        )
+    {
+        bool rebase = shouldRebase();
+        bool allocate = shouldAllocate();
+        bool isSuccessful = rebase || allocate;
+
+        ActionInfo memory rebaseInfo = ActionInfo({
+            actionType: REBASE,
+            run: rebase,
+            param: "foo"
+        });
+        ActionInfo memory allocateInfo = ActionInfo({
+            actionType: ALLOCATE,
+            run: allocate,
+            param: "foo"
+        });
+        KeepInfo memory keepInfo = KeepInfo({
+            rebaseInfo: rebaseInfo,
+            allocateInfo: allocateInfo
+        });
+
+        bytes memory data = abi.encode(keepInfo);
+        return (isSuccessful, data);
+    }
+
+    function performUpkeep(bytes calldata dynamicData) external {
+        KeepInfo memory keepInfo = abi.decode(dynamicData, (KeepInfo));
+
+        require(
+            keepInfo.rebaseInfo.run || keepInfo.allocateInfo.run,
+            "No keeper actions are enabled"
+        );
+        require(
```

Comment:
We could probably simplify this and just have performUpkeep check each task, and then execute if it needs to operate. This avoids having to trust user input, and avoids the overhead of passing information about what to execute out and back into the contract. As it is, an attacker could call "allocate" if a rebase was ready to execute. This wouldn't be an actual security problem since those are public anyway, but its aways worth being paranoid.

Reply:
I THINK I did what you suggested. All of my abi encode/decode work out the window! 😭  I'm all for simpler though...